### PR TITLE
CAT: set `additionalProperties` recursively for objects

### LINKED ISSUE #32191: CAT verifying the schema of all records matches the connector's schema
## What
[We're dropping fields because the schema of our connectors are not kept up to date](https://app.datadoghq.com/dashboard/hq4-ush-fht/connector-problems?refresh_mode=sliding&from_ts=1698443459784&to_ts=1699048259784&live=true).

We can implement a SAT that uses [the schema inferrer used in the connector builder](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py) to infer the schema of each record and ensure it matches the schema defined by the connector. 

This test should be part of the high-strictness test suite

### PR DESCRIPTION
**Pull request by artem1205**
## What

Resolve https://github.com/airbytehq/airbyte/issues/32191

## How

Set `additionalProperties` recursively for objects inside jsonschema.
Skip objects if no properties are mentioned.

## Recommended reading order
1. `airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/asserts.py`
2. `airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_asserts.py`

## 🚨 User Impact 🚨

no breaking changes

## Pre-merge Actions

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_asserts.py [hunk 1]
```diff
@@ -119,3 +119,82 @@ def test_validate_records_format(record, configured_catalog, valid):
...
+        "simple_schema",
+        "schema_with_object_without_properties",
+        "schema_with_object_with_properties",
+        "schema_with_array_of_strings",
+        "schema_with_array_of_objects",
+        "schema_with_array_of_objects_with_properties",
+    ],
+)
+def test_validate_additional_properties(original_schema, expected_schema):
+    assert _enforce_false_additional_properties(original_schema) == expected_schema
```
Comment (maxi297) [inline airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_asserts.py:200]: It is weird to me to test a private methods. There is nothing enforcing the public method `verify_records_schema` to use `_enforce_false_additional_properties` hence, the behavior of the public interface might be very different than what we expect given this suite of tests; we have no guarantee that we are testing that there are no additional properties.

Can we have the same tests setup but call `verify_records_schema` instead and test the public behavior?

Comment (artem1205) [reply airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_asserts.py]: test updated, please take a look

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/asserts.py [hunk 2]
```diff
@@ -67,7 +90,7 @@ def verify_records_schema(
     for stream in catalog.streams:
         schema_to_validate_against = stream.stream.json_schema
         if fail_on_extra_columns:
-            schema_to_validate_against = _enforce_no_additional_top_level_properties(schema_to_validate_against)
+            schema_to_validate_against = _enforce_false_additional_properties(schema_to_validate_against)
```
Comment (alafanechere) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/asserts.py:93]: I would find it more elegant to implement a custom [Validator](https://python-jsonschema.readthedocs.io/en/latest/api/jsonschema/protocols/#jsonschema.protocols.Validator) class which would enforce additional properties to false.

**Review by alafanechere — APPROVED** [review]
Minor suggestion. Please update the CHANGELOG.md file

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/asserts.py (hunk 2 above)
Comment (artem1205) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/asserts.py]: agreed, moved schema transformation inside validator class

**Pull request merged**